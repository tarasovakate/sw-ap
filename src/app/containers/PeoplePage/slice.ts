import { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from 'utils/@reduxjs/toolkit';
import { ContainerState, PeopleData, Film } from './types';

// The initial state of the PeoplePage container
export const initialState: ContainerState = {
  peopleData: {
    count: 0,
    results: [],
  },
  films: [],
  isFetchingPeopleAPI: false,
  isFetchingFilmAPI: false,
};

const peoplePageSlice = createSlice({
  name: 'peoplePage',
  initialState,
  reducers: {
    getPeople(state, action: PayloadAction<number>) {},
    setPeople(state, action: PayloadAction<PeopleData>) {
      state.peopleData = action.payload;
    },
    getFilms(state, action: PayloadAction<string[]>) {},
    setFilms(state, action: PayloadAction<Film[]>) {
      state.films = action.payload;
    },
    setFetchingPeopleAPIStatus(state, action: PayloadAction<boolean>) {
      state.isFetchingPeopleAPI = action.payload;
    },
    setFetchingFilmAPIStatus(state, action: PayloadAction<boolean>) {
      state.isFetchingFilmAPI = action.payload;
    },
  },
});

export const { actions, reducer, name: sliceKey } = peoplePageSlice;
