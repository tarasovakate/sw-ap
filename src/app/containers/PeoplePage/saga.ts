import { call, put, takeLatest, all } from 'redux-saga/effects';
import { request } from 'utils/request';
import { actions } from './slice';

const PEOPLE_API_URL = 'https://swapi.dev/api/people/';

export function* getPeople(action) {
  const { payload }: { payload: number } = action;
  yield put(actions.setFetchingPeopleAPIStatus(true));
  try {
    const result = yield call(request, `${PEOPLE_API_URL}?page=${payload}`);
    yield put(actions.setPeople(result));
  } catch (err) {
    throw err;
  } finally {
    yield put(actions.setFetchingPeopleAPIStatus(false));
  }
}

export function* getFilms(action) {
  const { payload }: { payload: string[] } = action;
  yield put(actions.setFetchingFilmAPIStatus(true));
  try {
    const result = yield all(payload.map(url => call(request, url)));
    yield put(actions.setFilms(result));
  } catch (err) {
    throw err;
  } finally {
    yield put(actions.setFetchingFilmAPIStatus(false));
  }
}

export function* peoplePageSaga() {
  yield takeLatest(actions.getPeople, getPeople);
  yield takeLatest(actions.getFilms, getFilms);
}
