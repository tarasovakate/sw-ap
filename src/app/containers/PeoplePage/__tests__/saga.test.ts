import { put, takeLatest } from 'redux-saga/effects';
import * as slice from '../slice';

import { peoplePageSaga, getPeople, getFilms } from '../saga';

describe('getRepos Saga', () => {
  const mockGetPeopleAction = { payload: 1 };
  const mockGetFilmAction = { payload: ['https://test-url.com'] };
  let getPeopleSagaIterator: ReturnType<typeof getPeople>;
  let getFilmsSagaIterator: ReturnType<typeof getFilms>;

  it('should set fetching people status at the beginning', () => {
    getPeopleSagaIterator = getPeople(mockGetPeopleAction);
    const putPeopleLoadingDescriptor = getPeopleSagaIterator.next().value;
    expect(putPeopleLoadingDescriptor).toEqual(
      put(slice.actions.setFetchingPeopleAPIStatus(true)),
    );
  });

  it('should set people data to store', () => {
    const mockResult = {
      count: 82,
      next: 'http://swapi.dev/api/people/?page=3',
      previous: 'http://swapi.dev/api/people/?page=1',
      results: [
        {
          name: 'Anakin Skywalker',
          height: '188',
          mass: '84',
          hair_color: 'blond',
          skin_color: 'fair',
          eye_color: 'blue',
          birth_year: '41.9BBY',
          gender: 'male',
          homeworld: 'http://swapi.dev/api/planets/1/',
          films: [
            'http://swapi.dev/api/films/4/',
            'http://swapi.dev/api/films/5/',
            'http://swapi.dev/api/films/6/',
          ],
          created: '2014-12-10T16:20:44.310000Z',
          edited: '2014-12-20T21:17:50.327000Z',
          url: 'http://swapi.dev/api/people/11/',
        },
      ],
    };
    getPeopleSagaIterator = getPeople(mockGetPeopleAction);
    const putPeopleLoadingDescriptor = getPeopleSagaIterator.next().value;
    expect(putPeopleLoadingDescriptor).toEqual(
      put(slice.actions.setFetchingPeopleAPIStatus(true)),
    );

    const requestDescriptor = getPeopleSagaIterator.next().value;
    expect(requestDescriptor).toMatchSnapshot();

    const setResultDescriptor = getPeopleSagaIterator.next(mockResult).value;
    expect(setResultDescriptor).toEqual(
      put(slice.actions.setPeople(mockResult)),
    );

    const disablePeopleLoadingDescriptor = getPeopleSagaIterator.next().value;
    expect(disablePeopleLoadingDescriptor).toEqual(
      put(slice.actions.setFetchingPeopleAPIStatus(false)),
    );
  });

  it('should set films data to store', () => {
    const mockResult = [
      {
        title: 'A New Hope',
        director: 'mock',
        producer: 'mock',
      },
    ];
    getFilmsSagaIterator = getFilms(mockGetFilmAction);
    const putFilmLoadingDescriptor = getFilmsSagaIterator.next().value;
    expect(putFilmLoadingDescriptor).toEqual(
      put(slice.actions.setFetchingFilmAPIStatus(true)),
    );

    const requestDescriptor = getFilmsSagaIterator.next().value;
    expect(requestDescriptor).toMatchSnapshot();

    const setResultDescriptor = getFilmsSagaIterator.next(mockResult).value;
    expect(setResultDescriptor).toEqual(
      put(slice.actions.setFilms(mockResult)),
    );

    const disableFilmLoadingDescriptor = getFilmsSagaIterator.next().value;
    expect(disableFilmLoadingDescriptor).toEqual(
      put(slice.actions.setFetchingFilmAPIStatus(false)),
    );
  });
});
