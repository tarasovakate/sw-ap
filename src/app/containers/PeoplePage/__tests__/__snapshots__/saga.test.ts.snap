// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRepos Saga should set films data to store 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": true,
  "payload": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "https://test-url.com",
        ],
        "context": null,
        "fn": [Function],
      },
      "type": "CALL",
    },
  ],
  "type": "ALL",
}
`;

exports[`getRepos Saga should set people data to store 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://swapi.dev/api/people/?page=1",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
